File: _21_Symmetric_Tree.ts
class TreeNode3 {
  val: number;
  left: TreeNode3 | null;
  right: TreeNode3 | null;

  constructor(val?: number, left?: TreeNode3 | null, right?: TreeNode3 | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// function isSymmetric(root: TreeNode3 | null): boolean {
//   if (!root) return true;

//   function isMirror(left: TreeNode3 | null, right: TreeNode3 | null): boolean {
//     if (!left && !right) return true;
//     if (!left || !right) return false;
//     if (left.val !== right.val) return false;

//     return isMirror(left.left, right.right) && isMirror(left.right, right.left);
//   }

//   return isMirror(root.left, root.right);
// }

// Other solution


function isSymmetric(root: TreeNode3 | null): boolean {
    if (!root) return true;

    function isMirror(t1: TreeNode3 | null, t2: TreeNode3 | null): boolean {
        if (!t1 && !t2) return true;
        if (!t1 || !t2) return false;

        return (
            t1.val === t2.val &&
            isMirror(t1.left, t2.right) &&
            isMirror(t1.right, t2.left)
        );
    }

    return isMirror(root.left, root.right);
}


function main() {
  const root1 = new TreeNode3(
    1,
    new TreeNode3(2, new TreeNode3(3), new TreeNode3(4)),
    new TreeNode3(2, new TreeNode3(4), new TreeNode3(3))
  );
  console.log("Ejemplo 1:", isSymmetric(root1));

  const root2 = new TreeNode3(
    1,
    new TreeNode3(2, null, new TreeNode3(3)),
    new TreeNode3(2, null, new TreeNode3(3))
  );
  console.log("Ejemplo 2:", isSymmetric(root2));

  const root3 = new TreeNode3(1); 
  console.log("Ejemplo 3:", isSymmetric(root3)); 

  const root4 = new TreeNode3(
    1,
    new TreeNode3(2, new TreeNode3(3), null),
    new TreeNode3(2, null, new TreeNode3(3))
  );
  console.log("Ejemplo 4:", isSymmetric(root4)); 
}

main();

File: _21_Symmetric_Tree.java
import java.util.LinkedList;
import java.util.Queue;

public class _21_Symmetric_Tree {
    public static void main(String[] args) {
        Solution sol = new Solution();

        TreeNode root1 = new TreeNode(1,
            new TreeNode(2, new TreeNode(3), new TreeNode(4)),
            new TreeNode(2, new TreeNode(4), new TreeNode(3))
        );
        System.out.println("Ejemplo 1: " + sol.isSymmetric(root1)); 

        TreeNode root2 = new TreeNode(1,
            new TreeNode(2, null, new TreeNode(3)),
            new TreeNode(2, null, new TreeNode(3))
        );
        System.out.println("Ejemplo 2: " + sol.isSymmetric(root2));
    }
}

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

//My solution
// class Solution {
//     public boolean isSymmetric(TreeNode root) {
//         return isMirror(root.left, root.right);
//     }

//     private boolean isMirror(TreeNode left, TreeNode right) {

//         if (left == null && right == null) return true;
        
//         if (left == null || right == null) return false;

//         if (left.val != right.val) return false;

//         return isMirror( left.left, right.right ) && isMirror(left.right, right.left);
//     }

// }

class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root.left);
        queue.add(root.right);

        while (!queue.isEmpty()) {
            TreeNode t1 = queue.poll();
            TreeNode t2 = queue.poll();

            if (t1 == null && t2 == null) continue;
            if (t1 == null || t2 == null) return false;
            if (t1.val != t2.val) return false;

            queue.add(t1.left);
            queue.add(t2.right);

            queue.add(t1.right);
            queue.add(t2.left);
        }

        return true;
    }
}